### DISCUSSION TITLE
Service Workers - Background Sync

### DISCUSSION THREAD

**Discussion by Coastline23** [anvilistas/anvil-labs#147] [Q&A]
Hi @s-cork,

You gave me an example app a while back to demonstrate how I might implement Background sync on a messaging app using Anvil Labs Service Workers [https://anvil.works/forum/t/js-anvil-call-error-anvil-is-not-defined/16968/4?u=apearce](url)

This appeared to work as required but I have not implemented it until now.

I have finally returned to this part of my app but I'm having issues with the example app.

It sometimes works:
 
1) storing the message in indexed db 
2) registering the sync
3) triggering the Service Worker  
4) calling the server function to save the messages in the apptable (I added code to do this part) 
5) clearing the message from indexed db.

But then, seemingly arbitrarily, subsequent messages just stop at the anvil server call 4).

app log:
```
<module 'anvil_labs.service_worker' from 'app/anvil_labs/service_worker.py'>
0.0.1
Warning: No method "text_area_1_change" in form Form1: cannot set the 'change' event handler of self.MessageText.
<SERVICE WORKER STDOUT>: Service Worker Loaded
<SERVICE WORKER READY>
Registering Sync Event
<SERVICE WORKER STDOUT>: sync fired
<SERVICE WORKER STDOUT>: Saving Messages: [{'id': '339818c8-57df-cfd7-84d5-28578b74eae4', 'tmref': 1234, 'message': 'mess1', 'sent': '2024-02-12 14:13:59'}]
Storing Mesage mess1
Storing Mesage mess2
Registering Sync Event
<SERVICE WORKER STDOUT>: sync fired
<SERVICE WORKER STDOUT>: Saving Messages: [{'id': '339818c8-57df-cfd7-84d5-28578b74eae4', 'tmref': 1234, 'message': 'mess1', 'sent': '2024-02-12 14:13:59'}, {'id': '456ae813-0995-906e-6f18-2f4352fbbec6', 'tmref': 1234, 'message': 'mess2', 'sent': '2024-02-12 14:14:29'}]
Storing Mesage mess3
Registering Sync Event
<SERVICE WORKER STDOUT>: sync fired
<SERVICE WORKER STDOUT>: Saving Messages: [{'id': '339818c8-57df-cfd7-84d5-28578b74eae4', 'tmref': 1234, 'message': 'mess1', 'sent': '2024-02-12 14:13:59'}, {'id': '456ae813-0995-906e-6f18-2f4352fbbec6', 'tmref': 1234, 'message': 'mess2', 'sent': '2024-02-12 14:14:29'}, {'id': 'e54dd95d-9b6e-30da-2aae-ee38dd246d93', 'tmref': 1234, 'message': 'mess3', 'sent': '2024-02-12 14:14:49'}]
```

processing appears to stop at that point.  it does not produce any console/log errors nor does it return processing to the remaining Service worker Code to clear messages from indexed db.  It just appears to stop dead at the anvil server call. 

The problem is intermittent but repetative.  I can run it successfully just once, or many times but then it just stops working.

It also threw a strange import error at one point:

```
<SERVICE WORKER STDERR>: ImportError('Attempted relative import in non-package')
ImportError: Attempted relative import in non-package
at sw, line 8
called from app/anvil_labs/service_worker.py, line 24
```

This was with no change to the code.  I restarted the app and the error was gone.

I'm stuck!  any ideas?

here's a clone [to](https://anvil.works/build#clone:UOOAMW5VNYDNBP33=HDDZB2CTB5QULQTWJNWNZJ4I|null|null) my slightly amended version of your example app. Just added a datetime to the message and code to store the message in the apptable.

Thanks

**Comment by s-cork** [comment]
thanks for reporting, I'll take a look at some point
